server/src/main/java/org/apache/iotdb/db/service/RegionMigrateService.java

```diff
@@ -474,4 +502,479 @@ private boolean isFailed(TSStatus status) {
+        } catch (Throwable e) {
+          addPeerSucceed = false;
+          taskLogger.error(
+              "add new peer {} for region {} error, retry times: {}", newPeerNode, regionId, i, e);
+          status.setCode(TSStatusCode.MIGRATE_REGION_ERROR.getStatusCode());
+          status.setMessage(
+              "add peer "
+                  + newPeerNode
+                  + " for region: "
+                  + regionId
+                  + " error, exception: "
+                  + e.getMessage());
+        }
+        if (addPeerSucceed && resp != null && resp.isSuccess()) {
+          break;
+        }
+      }
+      if (!addPeerSucceed || resp == null || !resp.isSuccess()) {
+        taskLogger.error(
+            "add new peer {} for region {} failed, resp: {}", newPeerNode, regionId, resp);
+        status.setCode(TSStatusCode.MIGRATE_REGION_ERROR.getStatusCode());
+        status.setMessage("add new peer " + newPeerNode + " for region " + regionId + "failed");
+        return status;
+      }
+
+      taskLogger.info("succeed to add peer {} for region {}", newPeerNode, regionId);
+      status.setCode(TSStatusCode.SUCCESS_STATUS.getStatusCode());
+      status.setMessage("add peer " + newPeerNode + " for region " + regionId + " succeed");
+      return status;
+    }
+
+    private ConsensusGenericResponse addRegionPeer(ConsensusGroupId regionId, Peer newPeer) {
+      ConsensusGenericResponse resp;
+      if (regionId instanceof DataRegionId) {
+        resp = DataRegionConsensusImpl.getInstance().addPeer(regionId, newPeer);
+      } else {
+        resp = SchemaRegionConsensusImpl.getInstance().addPeer(regionId, newPeer);
+      }
+      return resp;
+    }
+
+    private TEndPoint getConsensusEndPoint(
+        TDataNodeLocation nodeLocation, ConsensusGroupId regionId) {
+      if (regionId instanceof DataRegionId) {
+        return nodeLocation.getDataRegionConsensusEndPoint();
+      }
+      return nodeLocation.getSchemaRegionConsensusEndPoint();
+    }
+
+    private boolean isSucceed(TSStatus status) {
+      return status.getCode() == TSStatusCode.SUCCESS_STATUS.getStatusCode();
+    }
+
+    private boolean isFailed(TSStatus status) {
+      return !isSucceed(status);
+    }
+  }
+
+  private static class RemoveRegionPeerTask implements Runnable {
+    private static final Logger taskLogger = LoggerFactory.getLogger(RegionMigrateTask.class);
+
+    // migrate which region
+    private final TConsensusGroupId tRegionId;
+
+    // migrate from which node
+    private final TDataNodeLocation fromNode;
+
+    // transfer leader to which node
+    private final TDataNodeLocation newLeaderNode;
+
+    public RemoveRegionPeerTask(
+        TConsensusGroupId tRegionId, TDataNodeLocation fromNode, TDataNodeLocation newLeaderNode) {
+      this.tRegionId = tRegionId;
+      this.fromNode = fromNode;
+      this.newLeaderNode = newLeaderNode;
+    }
+
+    @Override
+    public void run() {
+      // changeLeader();
```

Comment:
I supposed that we'll never call changeLeader inside the process of RemoveRegionPeer?

Reply:
delete this.